src/sage_interface.h

```diff
@@ -0,0 +1,167 @@
+/* LibGAP - a shared library version of the GAP kernel
+ * Copyright (C) 2013 Volker Braun <vbraun.name@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA. 
+ */
+
+#ifndef LIBGAP__H
+#define LIBGAP__H
+
+
+/*************************************************************************/
+/*** Initialize / Finalize ***********************************************/
+/*************************************************************************/
+
+/* To setup libGAP, you must call libgap_initialize() to
+ * set argv and env
+ *
+ * Two mandatory component of argv is "-K", "memory_pool_size" to set the
+ * libgap memory pool size. For example,
+ *
+ * argc = 5
+ * argv = "libgap", "-l", "/path/to/gap-library", "-K", "64M", NULL
+ * env = NULL
+ */
+void libgap_initialize(int argc, char** argv);
+
+
+/* Currently, this does nothing. One day, it should correctly
+ * finalize libgap so that it can be reinitialized with a different
+ * memory pool...
+ */
+void libgap_finalize(void);
+
+
+/*************************************************************************/
+/*** Local (per-function) Initialization  ********************************/
+/*************************************************************************/
+
+/* You must call libgap_mark_stack_bottom() in every function that
+ * calls into the libGAP C functions. The reason is that the GAP
+ * memory manager will automatically keep objects alive that are
+ * referenced in local (stack-allocated) variables. While convenient,
+ * this requires to look through the stack to find anything that looks
+ * like an address to a memory bag.
+ *
+ * I suggest you use libgap_enter / libgap_exit (see below) instead of
+ * calling libgap_mark_stack_bottom manually.
+ */
+
+extern void* StackBottomBags;
+
+/* This is implemented as a macro to run in the same stack frame as
+ * the calling function */
+#ifdef __GNUC__
+#define libgap_mark_stack_bottom()              \
+  StackBottomBags = __builtin_frame_address(0);
+#else  /* try x86 asm */ 
+#define libgap_mark_stack_bottom()                 \
+  register void* ebp asm("ebp");                   \
+  StackBottomBags = ebp;
+#endif
```

Comment:
Relying on things like `__GNUC__` goes strictly against the autoconf philosophy, which mandates that one detect features, instead of blindly relying that certain global defines imply certain features. E.g. various non-GNU compilers #define __GNUC__ in order to trick code that thinks it's a good idea to use __GNUC__ to work -- but as a result, some code (like the above) may become broken. Also, ancient GNU C versions won't have `__builtin_frame_address` either.

So on the long term, we should have a test for the `__builtin_frame_address` builtin (see also the existing test for builtins). And another test to see if we are on x86 (or x86_64...) and then use the fallback code above -- and of course if neither is available, a more helpful message could be output.